build/source/engine/allocspace.f90

```diff
@@ -19,405 +19,348 @@
- implicit none
- ! dummy variables
- integer(i4b),intent(in)              :: nHRU        ! number of HRUs
- integer(i4b),intent(out)             :: err         ! error code
- character(*),intent(out)             :: message     ! error message
- ! local variables
- integer(i4b)                         :: iHRU        ! loop through HRUs
- integer(i4b)                         :: nVar        ! number of variables
- ! initialize errors
- err=0; message="alloc_time/"
- ! check that the metadata structure is allocated
- if(.not.associated(time_meta))then
-  err=10; message=trim(message)//"metadataNotInitialized"; return
- endif
- ! initialize top-level data structure
- if(associated(time_hru)) deallocate(time_hru)
- allocate(time_hru(nHRU),stat=err)
- if(err/=0)then; err=20; message=trim(message)//"problemAllocateDataTopLevel"; return; endif
- ! initialize second level data structure
- nVar = size(time_meta)
- do iHRU=1,nHRU
-  allocate(time_hru(iHRU)%var(nVar),stat=err)
-  if(err/=0)then; err=20; message=trim(message)//"problemAllocateData2ndLevel"; return; endif
-  ! set values to missing
-  time_hru(iHRU)%var(:) = missingInteger
- end do
- end subroutine alloc_time
+  ! ************************************************************************************************
+  ! internal subroutine allocGlobal: allocate space for global data structures 
+  ! ************************************************************************************************
+  subroutine allocGlobal(nHRU,metaStruct,dataStruct,err,message)
+  implicit none
+  ! input
+  integer(i4b),intent(in)         :: nHRU           ! number of HRUs
+  type(var_info),intent(in)       :: metaStruct(:)  ! metadata structure
+  ! output
+  class(*),intent(out)            :: dataStruct     ! data structure
+  integer(i4b),intent(out)        :: err            ! error code
+  character(*),intent(out)        :: message        ! error message
+  ! local variables
+  logical(lgt)                    :: check          ! .true. if structure is already allocated
+  integer(i4b)                    :: iHRU           ! loop htrough HRUs
+  character(len=256)              :: cmessage       ! error message of the downwind routine
+  ! initialize error control
+  err=0; message='allocGlobal/'
+  
+  ! initialize allocation check
+  check=.false.
+  
+  ! allocate HRU dimension
+  select type(dataStruct)
+   type is (spatial_int);       if(allocated(dataStruct%hru))then; check=.true.; else; allocate(dataStruct%hru(nHRU),stat=err); endif 
+   type is (spatial_intVec);    if(allocated(dataStruct%hru))then; check=.true.; else; allocate(dataStruct%hru(nHRU),stat=err); endif
+   type is (spatial_double);    if(allocated(dataStruct%hru))then; check=.true.; else; allocate(dataStruct%hru(nHRU),stat=err); endif
+   type is (spatial_doubleVec); if(allocated(dataStruct%hru))then; check=.true.; else; allocate(dataStruct%hru(nHRU),stat=err); endif
+  end select
```

Comment:
No. It is possible for the data type to be none of these.

However, for clarity, I should include a `class default` statement that does nothing, and include a comment that it is legal and acceptable to not be any of the specified cases.